src/main/java/com/ngs/stash/externalhooks/hook/ExternalHookScript.java

```diff
@@ -0,0 +1,256 @@
+import com.atlassian.bitbucket.hook.script.HookScriptType;
+import com.atlassian.bitbucket.hook.script.HookScriptUpdateRequest;
+import com.atlassian.bitbucket.permission.Permission;
+import com.atlassian.bitbucket.permission.PermissionService;
+import com.atlassian.bitbucket.scope.Scope;
+import com.atlassian.bitbucket.server.StorageService;
+import com.atlassian.bitbucket.setting.Settings;
+import com.atlassian.bitbucket.setting.SettingsValidationErrors;
+import com.atlassian.bitbucket.user.SecurityService;
+import com.atlassian.sal.api.pluginsettings.PluginSettings;
+import com.atlassian.sal.api.pluginsettings.PluginSettingsFactory;
+import com.atlassian.upm.api.license.PluginLicenseManager;
+import com.atlassian.upm.api.license.entity.PluginLicense;
+import com.atlassian.upm.api.util.Option;
+import com.google.common.escape.Escaper;
+import com.google.common.escape.Escapers;
+import org.apache.commons.io.FilenameUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import java.io.File;
+import java.util.Optional;
+
+public class ExternalHookScript {
+
+    public static final String PLUGIN_ID = "com.ngs.stash.externalhooks.external-hooks";
+
+    private final PluginLicenseManager pluginLicenseManager;
+
+    private static Logger log = LoggerFactory.getLogger(
+            ExternalHookScript.class.getSimpleName()
+    );
+
+    public final Escaper SHELL_ESCAPE;
+
+    private AuthenticationContext authCtx;
+    private PermissionService permissions;
+    private ClusterService clusterService;
+    private StorageService storageProperties;
+    private HookScriptService hookScriptService;
+    private PluginSettings pluginSettings;
+    private String hookComponentId;
+    private String hookId;
+    private HookScriptType hookScriptType;
+    private RepositoryHookTrigger repositoryHookTrigger;
+    private SecurityService securityService;
+
+    public ExternalHookScript(
+            AuthenticationContext authenticationContext,
+            PermissionService permissions,
+            PluginLicenseManager pluginLicenseManager,
+            ClusterService clusterService,
+            StorageService storageProperties,
+            HookScriptService hookScriptService,
+            PluginSettingsFactory pluginSettingsFactory,
+            SecurityService securityService,
+            String hookComponentId,
+            HookScriptType hookScriptType,
+            RepositoryHookTrigger repositoryHookTrigger) {
+
+        this.authCtx = authenticationContext;
+        this.permissions = permissions;
+        this.storageProperties = storageProperties;
+        this.pluginLicenseManager = pluginLicenseManager;
+        this.clusterService = clusterService;
+        this.hookScriptService = hookScriptService;
+        this.pluginSettings = pluginSettingsFactory.createGlobalSettings();
+        this.hookComponentId = hookComponentId;
+        this.hookId = PLUGIN_ID + ":" + hookComponentId;
```

Comment:
The way your `hookId` works isn't really safe for how you're trying to use it here. Effectively, your 3 hook classes exist for every project and repository (or, more exactly, they can be _configured_ for every project or repository). But you're only using, in essence, 3 `hookId` values--and they're the same for every project/repository.

Since you're using `PluginSettingsFactory.createGlobalSettings()`, when your 3 hook types are configured in different repositories, they're going to overwrite each other. You need to use the `Scope` object you're given to produce a unique plugin settings key, and your `deleteHookScriptByKey` needs to take the `Scope` from the `RepositoryHookEvent` in addition to the hook key so it can generate the right plugin settings key to check.

Should be pretty simple to confirm this doesn't work; just try to configure the same hook, like `ExternalPreReceiveHook`, in two different repositories.